### PR TITLE
Introduce hardware/bladeRF module

### PR DESCRIPTION
**Pull request by greydot**
###### Motivation for this change
Update libbladeRF to the latest version. Add module for easier system configuration.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by matthewbauer** [comment]
@grahamcofborg build libbladeRF

**Comment by markuskowa** [comment]
CC @lukeadams any ideas how to fix the darwin build?

**Comment by greydot** [comment]
@markuskowa it seems that malloc.h is a non-standard Linux header. It could be removed probably, but I don't have a Mac to test the result.

**Comment by matthewbauer** [comment]
It looks like they fixed it in https://github.com/Nuand/bladeRF/commit/0cb4ea888543b2dc75b876f7024e180854fbe9c3#diff-d5dd492f2296bbf8a94924c15c9efc09
You can probably add it as a patch.

File: pkgs/development/libraries/libbladeRF/default.nix [hunk 1]
```diff
@@ -21,6 +21,11 @@ stdenv.mkDerivation rec {
   # Fixup shebang
   prePatch = "patchShebangs host/utilities/bladeRF-cli/src/cmd/doc/generate.bash";
 
+  # Fixes macos and freebsd compilation issue.
+  # This is fixed upstream but hasn't been released yet.
+  # https://github.com/Nuand/bladeRF/commit/0cb4ea888543b2dc75b876f7024e180854fbe9c3
+  patches = ./fix-OSX-and-FreeBSD-compilation-issue.patch;
```
Comment (markuskowa) [inline pkgs/development/libraries/libbladeRF/default.nix]: It is better to use `fetchPatch` here instead of providing the file.

Comment (greydot) [reply pkgs/development/libraries/libbladeRF/default.nix]: Makes sense. Fixed.

**Comment by markuskowa** [comment]
@GrahamcOfBorg build libbladeRF

File: pkgs/development/libraries/libbladeRF/default.nix [hunk 2]
```diff
@@ -21,6 +21,15 @@ stdenv.mkDerivation rec {
...
+  # https://github.com/Nuand/bladeRF/commit/0cb4ea888543b2dc75b876f7024e180854fbe9c3
+  patches = [ (fetchpatch {
+                name = "fix-OSX-and-FreeBSD-build.patch";
+                url = "https://github.com/Nuand/bladeRF/commit/0cb4ea88.diff";
+                sha256 = "1ccpa69vz2nlpdnxprh4rd1pgphk82z5lfmbrfdkn7srw6nxl469";
+              })
+            ];
+
   # Let us avoid nettools as a dependency.
   patchPhase = ''
```
Comment (markuskowa) [inline pkgs/development/libraries/libbladeRF/default.nix]: This one overrides the patch phase. Thus the patch is never really applied. ``postPatch`` is the better choice here:
```suggestion
  postPatch= ''
```

Comment (Ekleog) [reply pkgs/development/libraries/libbladeRF/default.nix]: @greydot Could you handle this comment? It should (hopefully) fix the darwin build :)

Comment (greydot) [reply pkgs/development/libraries/libbladeRF/default.nix]: @Ekleog done. Sorry for the delay.

**Pull request merged**